src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java

```diff
@@ -62,6 +86,91 @@
     return caseData;
   }
 
+  @Override
+  public CaseDTO modifyAddress(final UUID caseId, final AddressChangeDTO addressChanges)
+      throws CTPException {
+
+    Optional<CollectionCase> caseMatch = dataRepo.readCollectionCase(caseId.toString());
+
+    if (!caseMatch.isPresent()) {
+      log.with("caseId", caseId).error("Failed to retrieve Case from storage");
+      throw new CTPException(CTPException.Fault.RESOURCE_NOT_FOUND, "Failed to retrieve Case");
+    }
+
+    CollectionCase rmCase = caseMatch.get();
+
+    CaseDTO caseData = createModifiedAddressCaseDetails(caseId, rmCase, addressChanges);
+
+    AddressModified originalAddress = addressMapperFacade.map(rmCase.getAddress());
+
+    AddressModified updatedAddress =
+        addressDTOMapperFacade.map(
+            addressChanges.getAddress(), addressMapperFacade.map(rmCase.getAddress()));
+
+    sendAddressModifiedEvent(caseId, originalAddress, updatedAddress);
+
+    return caseData;
+  }
+
+  /**
+   * Create case details with updated Address
+   *
+   * @param caseId requested caseId for which to update address
+   * @param rmCase original Case from repository
+   * @param addressChanges Changed address details from request
+   * @return CaseDTO updated case details
+   * @throws CTPException UPRN of stored address and request change do not match
+   */
+  private CaseDTO createModifiedAddressCaseDetails(
+      UUID caseId, CollectionCase rmCase, AddressChangeDTO addressChanges) throws CTPException {
+
+    CaseDTO caseData = mapperFacade.map(rmCase, CaseDTO.class);
+    if (!caseData.getAddress().getUprn().equals(addressChanges.getAddress().getUprn())) {
```

Comment:
The are-the-uprns equal check would be better done by the caller: modifyAddress(). 
Validating that the uprn for the retrieved case and that of the new address are equal doesn't sound like the responsibility of a method which is creating address case details.

Reply:
True, this bothered me a bit as well. The advantage was that after I had used the mapper to create the CaseDTO  I could compare both UPRN values as the custom Jackson converter UniquePropertyReferenceNumber class, simply annotating this class with the lombok EqualsAndHashCode annotation. Otherwise with all the different address types we have I had to compare UniquePropertyReferenceNumber with a String which seemed a faff, either creating an instance from the String or implementing an unnecessary toString on UniquePropertyReferenceNumber. On balance putting this test here seemed better?